### ISSUE TITLE
Client Side IMA SDK integration for ad tag with dash mediafile

### ISSUE THREAD

**Issue by DDeis** [google/shaka-player#2460]
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
I haven't found anything relating to my problem here or on ima-sdk documentation and forum so I'm trying my luck here.

**What version of Shaka Player are you using?**
nightly build

**Please ask your question**
Hello, I'm trying to integrate IMA SDK with shaka-player but ads with "application/dash+xml" mediafile cannot play with error 400. Checking with adsManager.isCustomPlaybackUsed(), shaka-player does not seem to be used to play ads.

I tried with my own implementation and then with shaka nightly build demo with the same result.
You can checkout the branch reproducing my issue: https://github.com/DDeis/shaka-player/tree/ima-dash-test
Running the demo with custom contents:
Manifest Url: `http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4`
Ad tag uri: `/demo/resources/vast-with-dash.xml`

The only changes I've made is switching to ima3_debug.js and using adRequest.adsResponse instead of adTagUri.

Do you have any idea on how to make it work ?

**Comment by TheModMaker** [comment]
If you are getting a 400 error, that is from the server for an unauthorized request.  Check your server and the URL to make sure the request is allowed.

@ismena Could you help with this?

**Comment by ismena** [comment]
Apologies, this issue has escaped my attention somehow.
I will take a look later this week.

**Comment by ismena** [comment]
Ok, what I'm seeing is that IMA is trying to request your ad tag and their GET request is failing due to CORS restrictions .
The exact error:
`Access to XMLHttpRequest at 'https://github.com/DDeis/shaka-player/blob/ima-dash-test/demo/resources/vast-with-dash.xml' from origin 'https://imasdk.googleapis.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.`

The request I used to get the error:
```
var adM = shakaDemoMain.player_.getAdManager();
var video = shakaDemoMain.video_;
var adContainer = video.ui.getControls().getAdContainer();
adM.initClientSide(adContainer, video);
var adsRequest = new google.ima.AdsRequest();
adsRequest.adTagUrl = 'https://github.com/DDeis/shaka-player/blob/ima-dash-test/demo/resources/vast-with-dash.xml';
adM.requestClientSideAds(adsRequest);
```

Were you just trying out our ad logic? What is your normal flow for this - where are your ad tags hosted?

The good thing is, our current bug cop is an expert CORS troubleshooter. @joeyparrish would you have advice for @DDeis ? Would disabling cors locally help if the request is coming from IMA?

**Comment by joeyparrish** [comment]
Disabling CORS locally could help.  In Chrome, that requires a command-line flag and separate user-data directory (separate profile).  In Safari, there's an option in the developer menu, if I recall correctly.

But CORS issues are common, especially when people try something in our demo for the first time.  It's so common, we have a label for it in GitHub.  :-)

The shortest explanation is that there is a security feature built into browsers that requires the server to send a certain header allowing access to that resource from a different origin (essentially the domain).  If your server doesn't explicitly allow the demo to access the ad tags, then the browser will reject the response and fail the request.

MDN has a wonderful in-depth article on CORS: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS

If you want to skip straight to the easiest answer, you can allow access from all origins by adding the following HTTP header to the response:

```
Access-Control-Allow-Origin: *
```

If you don't control the headers your server sends out, you will not be able to access the ad tag cross-origin.  You can always host a copy of our demo on the same origin or another origin allowed by your server to work around this.

**Comment by DDeis** [comment]
Thank you for your feedbacks and apologies for the late response, with covid-19 the project was stopped.

As for CORS, I don't have access to the ads server and the team who is in charge of it does not want to make a change. But actually my issue is not CORS related, I managed to bypass CORS by "manually" fetching the VAST manifest and then use it as the adResponse. Based @ismena request, it would be:

```js
var adM = shakaDemoMain.player_.getAdManager();
var video = shakaDemoMain.video_;
var adContainer = video.ui.getControls().getAdContainer();
adM.initClientSide(adContainer, video);
var adsRequest = new google.ima.AdsRequest();

// Changes below
var adTagUrl = 'https://github.com/DDeis/shaka-player/blob/ima-dash-test/demo/resources/vast-with-dash.xml';
adRequest.adsResponse = await fetch(adTagUri).then((response) => response.text());

adM.requestClientSideAds(adsRequest);
```

(You can see the same change on my test repo at [demo/main.js l1191](https://github.com/DDeis/shaka-player/commit/a8f25c6c7ad5a87a56ebf63a13869805141c3576#diff-3b4cc93c41d4067694a6d9ff6ae923efR1191)).

This way, I can play mp4 ads normally, but with DASH mediaFiles ads cue points are displayed properly but the ad closes immediately with `AdError 400: There was an error playing the video ad. Caused by: Error: There was an error playing the video ad.`.

From IMA documentation:

> ### 400 - General linear error
> **Likely cause of a 400 VAST error**
> This fatal error can be caused when a video player is unable to display the linear ad. It may
occur when the MediaFile is not a valid video file of the specified format, or because the browser restricted autoplay with sound, or another unknown reason.
>
> **Recommended action:**
> _Publisher_: Check for player issues, enable your player to detect for restricted autoplay, or alert the buyer.
> _Buyer_: Ensure that the VAST response returns a valid creative.

I don't think it's a problem with the VAST response nor autoplay related given that mp4 ads play correctly and that another team who is developping the same app on mobile with ExoPlayer for Android and IMA sdk using the same process (manually fetching VAST manifest to avoid CORS) are playing dash ads without problem.

So my guess is that the native html5 video player is used instead of shaka-player for playing ads, and my guess seems to be confirmed as `adsManager.isCustomPlaybackUsed()`is falsy when I log it (for instance in client_side_ad_manager.js `onAdStart_()` function).

I also tried to manually setting mime types (in client_side_ad_manager.js - onAdsManagerLoaded_()) with no luck:
```js
const adsRenderingSettings = new google.ima.AdsRenderingSettings();
adsRenderingSettings.mimeTypes = ['application/dash+xml', 'video/mp4'];
this.imaAdsManager_.updateAdsRenderingSettings(adsRenderingSettings);
```

Do you agree with my reasoning ? I'm fairly novice with IMA sdk and have basic understanding of shaka-player so maybe there are some things I'm not aware of.

**Comment by ismena** [comment]
> So my guess is that the native html5 video player is used instead of shaka-player for playing ads
Yes, you are right! The way Client Side HTML5 IMA works is by creating another MediaElement on top of what the client has, pausing our playback, playing ads in their element, then removing it and giving the control back to us to resume our (main content) playback.

Sounds like the IMA team might be better equipped to comment on your issue as I don't think it's with Shaka. I don't know IMA _that_ deeply, but I could reach out to the IMA team and ask what the best avenue is for asking their advice (where to send/submit an issue etc.)

Would that help?

**Comment by DDeis** [comment]
All right, I though Shaka would also be used in the overlaid player but I obviously got misled.

As I understand I should ask the IMA team for advice about using a custom ad playback like they did for Android. I know IMA sdk has a Google Group, actually I posted a message on it before opening this issue but their answer got me thinking it was maybe due to my Shaka integration.

Thank you again for your time, I let you know if I find a solution !

**Comment by ismena** [comment]
Sounds good! Sorry, I wish were able to be of more help :|
I'm closing this as external for now, please feel free to reopen or file a new issue if you need anything else.
